File: src/DashSpec.Core/Parsing/DiagramParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class DiagramParser
{
    private static readonly HashSet<string> BannedTooltipPropertyKeys = new(StringComparer.OrdinalIgnoreCase)
    {
        "tooltip",
        "tooltip_time",
        "tooltip_format",
        "tooltip_split",
        "tooltip_as",
    };

    public static DiagramDefinition Parse(TokenReader reader)
    {
        var name = reader.ReadIdent();
        return ParseAfterKindIdent(reader, name);
    }

    public static DiagramDefinition ParseAfterKindIdent(TokenReader reader, string name)
    {
        if (DiagramKindRegistry.TryResolve(name, out var spec))
        {
            var properties = PropertyBlockParser.Parse(
                reader,
                DiagramKindRegistry.GetProperties(name),
                $"diagram {name}",
                spec.AllowExtensionProperties);
            RejectLegacyTooltipProperties(properties, name);
            return new DiagramDefinition(name, properties);
        }

        var overrides = !reader.IsOnNewline() && !reader.IsEof
            ? PropertyBlockParser.Parse(
                reader,
                DiagramKindRegistry.AllBindingProperties(),
                $"diagram {name}",
                allowExtensionProperties: true)
            : new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        RejectLegacyTooltipProperties(overrides, name);
        return new DiagramDefinition(string.Empty, overrides, name);
    }

    private static void RejectLegacyTooltipProperties(
        IReadOnlyDictionary<string, string> properties,
        string context)
    {
        foreach (var key in properties.Keys)
        {
            if (BannedTooltipPropertyKeys.Contains(key))
            {
                throw new DashSpecParseException(
                    $"Diagram '{context}': property '{key}' was removed (ADR-0029). " +
                    "Use @tooltip entity and inspect block instead.");
            }
        }
    }
}

File: src/DashSpec.Modeling.Parse/Diagram/DiagramParser.fs
namespace DashSpec.Modeling.Parse.Diagram

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module DiagramParser =

    let private bannedTooltipPropertyKeys =
        set [ "tooltip"; "tooltip_time"; "tooltip_format"; "tooltip_split"; "tooltip_as" ]

    let private rejectLegacyTooltipProperties (properties: IReadOnlyDictionary<string, string>) context =
        for key in properties.Keys do
            if bannedTooltipPropertyKeys.Contains key then
                raise (DashSpecParseException($"Diagram '{context}': property '{key}' was removed (ADR-0029). Use @tooltip entity and inspect block instead."))

    let parseAfterKindIdent (reader: TokenReader) (name: string) =
        match DiagramKindRegistry.tryResolve name with
        | true, spec ->
            let properties =
                PropertyBlockParser.parse
                    reader
                    (DiagramKindRegistry.getProperties name)
                    $"diagram {name}"
                    spec.AllowExtensionProperties
                    false
            rejectLegacyTooltipProperties properties name
            { Kind = name; Properties = properties :> IReadOnlyDictionary<_, _>; UsePreset = None }
        | false, _ ->
            let overrides =
                if reader.IsOnNewline() || reader.IsEof then
                    Dictionary<string, string>(StringComparer.OrdinalIgnoreCase) :> IReadOnlyDictionary<_, _>
                else
                    PropertyBlockParser.parse
                        reader
                        (DiagramKindRegistry.allBindingProperties ())
                        $"diagram {name}"
                        true
                        false
            rejectLegacyTooltipProperties overrides name
            { Kind = ""; Properties = overrides; UsePreset = Some name }

    let parse (reader: TokenReader) =
        let name = reader.ReadIdent()
        parseAfterKindIdent reader name
